edn/src/query.rs

```diff
@@ -0,0 +1,1253 @@
+            ::SpannedValue::Text(ref x) =>
+                // TODO: intern strings. #398.
+                Some(PatternValuePlace::Constant(x.clone().into())),
+            ::SpannedValue::Uuid(ref u) =>
+                Some(PatternValuePlace::Constant(NonIntegerConstant::Uuid(u.clone()))),
+
+            // These don't appear in queries.
+            ::SpannedValue::Nil => None,
+            ::SpannedValue::NamespacedSymbol(_) => None,
+            ::SpannedValue::Keyword(_) => None,                // … yet.
+            ::SpannedValue::Map(_) => None,
+            ::SpannedValue::List(_) => None,
+            ::SpannedValue::Set(_) => None,
+            ::SpannedValue::Vector(_) => None,
+        }
+    }
+}
+
+impl PatternValuePlace {
+    // I think we'll want move variants, so let's leave these here for now.
+    #[allow(dead_code)]
+    fn into_pattern_non_value_place(self) -> Option<PatternNonValuePlace> {
+        match self {
+            PatternValuePlace::Placeholder       => Some(PatternNonValuePlace::Placeholder),
+            PatternValuePlace::Variable(x)       => Some(PatternNonValuePlace::Variable(x)),
+            PatternValuePlace::EntidOrInteger(x) => if x >= 0 {
+                Some(PatternNonValuePlace::Entid(x))
+            } else {
+                None
+            },
+            PatternValuePlace::IdentOrKeyword(x) => Some(PatternNonValuePlace::Ident(x)),
+            PatternValuePlace::Constant(_)       => None,
+        }
+    }
+
+    fn to_pattern_non_value_place(&self) -> Option<PatternNonValuePlace> {
+        match *self {
+            PatternValuePlace::Placeholder           => Some(PatternNonValuePlace::Placeholder),
+            PatternValuePlace::Variable(ref x)       => Some(PatternNonValuePlace::Variable(x.clone())),
+            PatternValuePlace::EntidOrInteger(x)     => if x >= 0 {
+                Some(PatternNonValuePlace::Entid(x))
+            } else {
+                None
+            },
+            PatternValuePlace::IdentOrKeyword(ref x) => Some(PatternNonValuePlace::Ident(x.clone())),
+            PatternValuePlace::Constant(_)           => None,
+        }
+    }
+}
+
+// Not yet used.
+// pub enum PullDefaultValue {
+//     EntidOrInteger(i64),
+//     IdentOrKeyword(Rc<Keyword>),
+//     Constant(NonIntegerConstant),
+// }
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PullConcreteAttribute {
+    Ident(Rc<Keyword>),
+    Entid(i64),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct NamedPullAttribute {
+    pub attribute: PullConcreteAttribute,
+    pub alias: Option<Rc<Keyword>>,
+}
+
+impl From<PullConcreteAttribute> for NamedPullAttribute {
+    fn from(a: PullConcreteAttribute) -> Self {
+        NamedPullAttribute {
+            attribute: a,
+            alias: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PullAttributeSpec {
+    Wildcard,
+    Attribute(NamedPullAttribute),
+    // PullMapSpec(Vec<…>),
```

Comment:
Cleanup

Reply:
These are specifically placeholders for the future, so I'm going to leave it.